### PR TITLE
i21: version of state() that allows filtering by index

### LINKED ISSUE #21: Version of state() method that accepts an index
This way we can do a particle-filter-with-trace approach more efficiently

### PR DESCRIPTION
**Pull request by richfitz**
This PR introduces a version of `state()` that filters by index so that we can return less than the full index. This is related to but different to the work required for #9, but will be needed for work in mcstate.

Partial fix for #33 (via a test) but this is something to think about in future.
Fixes #21

### CODE REVIEW

**Comment by johnlees** [comment]
Thoughts on making this v0.1 once done? This seems like everything that's needed for an interface through to mcstate & sircovid

**Comment by richfitz** [comment]
There's a couple more major bits of interface work, but I could bundle the worst of it into here perhaps. Specifically - there is no way of setting the index_y property on the object (#8) and no way of setting the initial state (#6). I've worked out how these need to go (it's complicated by not throwing exceptions and the state being determined by the data) and am sketching the interface for tha

**Comment by johnlees** [comment]
Ok! After those sounds sensible, then I will return to this in mcstate. Is there anything I can help with in the meantime?

**Comment by richfitz** [comment]
Let's get this reviewed, then I should be able to get the rest of the interface in today if all goes well

File: inst/include/dust/interface.hpp [hunk 1]
```diff
@@ -74,6 +84,20 @@ SEXP dust_state(SEXP ptr) {
   return Rcpp::NumericMatrix(n_state_full, n_particles, dat.begin());
 }
 
+template <typename T>
+SEXP dust_state_filter(Dust<T> *obj, Rcpp::IntegerVector r_index) {
```
[the remark this answers is not in the document]
Comment (richfitz) [reply inst/include/dust/interface.hpp]: After discussion on teams, trying "select" here

**Pull request merged**